### PR TITLE
Add shellcheck action, fix issues

### PR DESCRIPTION
**Pull request by matthiasbeyer**
This PR adds a `shellcheck` job. Shellcheck is reasonably fast that this does not impact our CI time _at all_.

This PR also fixes _most_ issues shellcheck reports. The ones I did not fix here are issues where I don't know how to go forward and might need some investigation from someone who knows better.

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Ah, we should probably fix the remaining ones in this PR as well.

Right now there are [23 issues left](https://github.com/thin-edge/thin-edge.io/runs/7703188972?check_suite_focus=true#step:3:169) that we need to fix before the job succeeds.

File: get-thin-edge_io.sh [hunk 1]
```diff
@@ -121,7 +121,7 @@ minimal) install_basic_components ;;
 full)
     install_basic_components
     install_tedge_agent
-    if apt -v &>/dev/null; then
+    if apt -v >/dev/null 2>&1; then
```
Comment (rina23q) [inline get-thin-edge_io.sh:124]: Are they the same?

File: get-thin-edge_io.sh [hunk 2]
```diff
@@ -134,8 +134,8 @@ esac
 rm -R /tmp/tedge
 
 # Test if tedge command is there and working
-tedge help >/dev/null
-if [ $? -eq 0 ]; then
+if ! tedge help >/dev/null;
+then
```
Comment (rina23q) [inline get-thin-edge_io.sh:138]: I am pretty sure that these are not equal change. Actually, after I ran the script in this PR, it went to the `else` case, even though the original version went to `then` case. We have to fix here.

File: plugins/tedge_docker_plugin/tedge_docker_plugin.sh [hunk 3]
```diff
@@ -15,8 +15,8 @@ EOF
 }
 
 unsupported_args_check() {
-    if ! [ -z $1 ]; then
-        echo "Unsupported arguments: $@"
+    if [ -n "$1" ]; then
+        echo "Unsupported arguments: $*"
```
Comment (rina23q) [inline plugins/tedge_docker_plugin/tedge_docker_plugin.sh:19]: What is this change?

**Review by rina23q — CHANGES_REQUESTED** [review]
I am happy to have `shellcheck` job in the pull request workflow.
 
However, I saw the behaviour change on `get-thin-edge_io.sh` script, which is blocker.
Others might work (I tested only the installation script, others not), but we need to check them before merging.

File: .github/workflows/pull-request-checks.yml [hunk 4]
```diff
@@ -303,6 +303,15 @@ jobs:
           command: test
           args: --release --no-run --target=armv7-unknown-linux-gnueabihf --all-features
 
+  shellcheck:
+    name: Shellcheck
+    needs: block-fixup
```
[the remark this answers is not in the document]
Comment (matthiasbeyer) [reply .github/workflows/pull-request-checks.yml:308]: That's right! I guess we can safely run this in parallel to everything else! Thanks for catching this!

File: get-thin-edge_io.sh (hunk 1 above)
Comment (matthiasbeyer) [reply get-thin-edge_io.sh:124]: Yes, those are equal. See [here](https://www.shellcheck.net/wiki/SC3020)!

File: get-thin-edge_io.sh (hunk 2 above)
Comment (matthiasbeyer) [reply get-thin-edge_io.sh:138]: Very good catch, thanks!

File: plugins/tedge_docker_plugin/tedge_docker_plugin.sh (hunk 3 above)
Comment (matthiasbeyer) [reply plugins/tedge_docker_plugin/tedge_docker_plugin.sh:19]: This is [SC2145](https://www.shellcheck.net/wiki/SC2145). In short: `$@` is array access in bash, `$*` is "all arguments as a string".

**Review by rina23q — APPROVED** [review]
LGTM.

The job is still failing [here](https://github.com/thin-edge/thin-edge.io/runs/7740278099?check_suite_focus=true). Does it mean, we need to fix more scripts later?

**Comment by matthiasbeyer** [comment]
That does mean we need to fix things _now_ because this PR is not ready yet, there are still warnings and errors in shell scripts.

File: get-thin-edge_io.sh [hunk 5]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
Comment (matthiasbeyer) [inline get-thin-edge_io.sh:1]: Someone that is not me must decide whether this is okay.

**Comment by matthiasbeyer** [comment]
Can someone please tell me what to do about 

* [SC2153](https://www.shellcheck.net/wiki/SC2153) in [`configure_bridge`](https://github.com/thin-edge/thin-edge.io/runs/7786319190?check_suite_focus=true#step:3:169)
* [SC1090](https://www.shellcheck.net/wiki/SC2153) in [multiple locations](https://github.com/thin-edge/thin-edge.io/runs/7786319190?check_suite_focus=true#step:3:170)
* [SC1091](https://www.shellcheck.net/wiki/SC2153) in [multiple locations](https://github.com/thin-edge/thin-edge.io/runs/7786319190?check_suite_focus=true#step:3:171)

**Comment by matthiasbeyer** [comment]
I am going to file issues for the things that are ignored in [5be1243](https://github.com/thin-edge/thin-edge.io/pull/1326/commits/5be1243a0410684316ff8791e78c5c5b091ca43e), [4e52a28](https://github.com/thin-edge/thin-edge.io/pull/1326/commits/4e52a283efbc078406b8523286725696636d03ba) and [ea85c68](https://github.com/thin-edge/thin-edge.io/pull/1326/commits/ea85c683f7a19a693afc8863f2fb884e5b05e375) after this is merged.

File: get-thin-edge_io.sh (hunk 5 above)
Comment (didier-wenzek) [reply get-thin-edge_io.sh:1]: For a quick install script `sh` would make more sense than `bash`.

What is specific to `bash` in this script? I noticed that you properly replaced `if apt -v &>/dev/null` by `apt -v >/dev/null 2>&1`. So `sh` should work.

Comment (matthiasbeyer) [reply get-thin-edge_io.sh:1]: > What is specific to bash in this script?

The escape sequences that are printed with `echo -e`. The `-e` flag is not specified by POSIX.

File: get-thin-edge_io.sh [hunk 6]
```diff
@@ -81,14 +81,14 @@ if [ -t 1 ]; then
     fi
 fi
 
-echo "${BLUE}Thank you for trying thin-edge.io! ${COLORRESET}\n"
+echo -e "${BLUE}Thank you for trying thin-edge.io! ${COLORRESET}\n"
```
Comment (didier-wenzek) [inline get-thin-edge_io.sh:84]: In order to be `sh` compatible.

```suggestion
echo "${BLUE}Thank you for trying thin-edge.io! ${COLORRESET}"
echo
```

File: get-thin-edge_io.sh [hunk 7]
```diff
@@ -134,11 +134,11 @@ esac
 rm -R /tmp/tedge
 
 # Test if tedge command is there and working
-tedge help >/dev/null
-if [ $? -eq 0 ]; then
-    echo "\n${BLUE}thin-edge.io is now installed on your system!${COLORRESET}"
+if tedge help >/dev/null;
+then
+    echo -e "\n${BLUE}thin-edge.io is now installed on your system!${COLORRESET}"
```
Comment (didier-wenzek) [inline get-thin-edge_io.sh:139]: In order to be `sh` compatible.

```suggestion
    echo
    echo "${BLUE}thin-edge.io is now installed on your system!${COLORRESET}"
```

File: get-thin-edge_io.sh [hunk 8]
```diff
@@ -134,11 +134,11 @@ esac
...
     echo ""
-    echo "You can go to our documentation to find next steps:${BLUE} https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/howto-guides/003_registration.md ${COLORRESET}"
+    echo -e "You can go to our documentation to find next steps:${BLUE} https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/howto-guides/003_registration.md ${COLORRESET}"
 else
-    echo "Something went wrong in the installation process please try the manual installation steps instead:\n https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/howto-guides/002_installation.md"
+    echo -e "Something went wrong in the installation process please try the manual installation steps instead:\n https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/howto-guides/002_installation.md"
```
Comment (didier-wenzek) [inline get-thin-edge_io.sh:143]: In order to be `sh` compatible.

```suggestion
    echo "Something went wrong in the installation process please try the manual installation steps instead:"
    echo "     https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/howto-guides/002_installation.md"
```

File: get-thin-edge_io.sh (hunk 5 above)
Comment (didier-wenzek) [reply get-thin-edge_io.sh:1]: In that case, I prefer to avoid the `echo -e` and use `sh`.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
All the `echo -e` can be removed in `get-thin-edge_io.sh`. So `sh` can be used.

File: get-thin-edge_io.sh (hunk 6 above)
Comment (matthiasbeyer) [reply get-thin-edge_io.sh:84]: Colors are also escape sequences.

Comment (didier-wenzek) [reply get-thin-edge_io.sh:84]: But do we really care about colors for a quick install script? Portability matters more.

Comment (matthiasbeyer) [reply get-thin-edge_io.sh:84]: I cannot decide that, but if you think I can remove the colors, I'm happy to rewrite this portion of the patchset to make sure we use `/bin/sh` and only features that are there according to `shellcheck` (thus removing colors)! :+1:

File: ci/configure_bridge.sh [hunk 9]
```diff
@@ -1,5 +1,15 @@
...
 
+if [ -z "$C8YDEVICE" ]; then
+    echo "C8YDEVICE is not set"
+    exit 1
+fi
+
+if [ -z "$C8YUSERNAME" ]; then
+    echo "C8YUSERNAME is not set"
+    exit 1
+fi
```
[the remark this answers is not in the document]
Comment (matthiasbeyer) [reply ci/configure_bridge.sh]: No problem! Just pushed a squash commit for this.

**Review by didier-wenzek — APPROVED** [review]
Approved.

**Comment by matthiasbeyer** [comment]
Squashed. :tada:

**Pull request merged**